src/hotspot/share/opto/rangeinference.hpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_OPTO_RANGEINFERENCE_HPP
+#define SHARE_OPTO_RANGEINFERENCE_HPP
+
+#include "utilities/globalDefinitions.hpp"
+#include <type_traits>
+
+class outputStream;
+class Type;
+class TypeInt;
+class TypeLong;
+
+template <class T>
+class RangeInt {
+public:
+  T _lo;
+  T _hi;
+};
+
+/**
+ * Bits that are known to be 0 or 1. A value v satisfies this constraint iff
+ * (v & zeros) == 0 && (v & ones) == ones. I.e, all bits that is set in zeros
+ * must be unset in v, and all bits that is set in ones must be set in v.
+ *
+ * E.g:
+ * zeros: 00110100
+ * ones:  10000010
+ * Then:  10001010 would satisfy the bit constraints
+ * while: 10011000 would not since the bit at the 4th position violates
+ * zeros and the bit at the 7th position violates ones
+ *
+ * A KnownBits is sane if there is no position at which a bit must be both set
+ * and unset at the same time. That is (zeros & ones) == 0.
+ */
+template <class U>
+class KnownBits {
+  static_assert(std::is_unsigned<U>::value, "bit info should be unsigned");
+
+public:
+  U _zeros;
+  U _ones;
+
+  bool is_satisfied_by(U v) const {
+    return (v & _zeros) == 0 && (v & _ones) == _ones;
+  }
+};
```

Comment:
It will be good if we add basic operations to KnowBits like.
KnownBits.getMaxValue() returning ~ZEROS
KnownBits.getMinValue() returning  ONE
KnownBits.and(KnownBits arg)
KnownBits.or(KnownBits arg)
KnownBits.xor(KnownBits args)
KnownBits.not()


These can be quite handy during data flow analysis using KnownBits

Reply:
Yes I think they would be helpful in later patches when implementing `Value` methods of several nodes to take advantage of additional `TypeInt` information.